server/routes/api/place.py

```diff
@@ -130,6 +130,67 @@ def api_name():
     return Response(json.dumps(result), 200, mimetype='application/json')
 
 
+@cache.memoize(timeout=3600 * 24)  # Cache for one day.
+def cached_i18n_name(dcids, locale):
+    """Returns localination names for set of dcids.
+
+    Args:
+        dcids: ^ separated string of dcids. It must be a single string for the cache.
+        locale: the desired localization language code.
+
+    Returns:
+        A dictionary of place international names, keyed by dcid.
+    """
+    dcids = dcids.split('^')
+    response = fetch_data('/node/property-values', {
+        'dcids': dcids,
+        'property': 'nameWithLanguage',
+        'direction': 'out'
+    },
+                          compress=False,
+                          post=True)
+    result = {}
+    for dcid in dcids:
+        values = response[dcid].get('out')
+        name_in_en = ''
+        result[dcid] = ''
+        for entry in values:
+            if entry['value'].endswith(locale):
+                i18nname_end_index = len(entry['value']) - len(locale) - 1
+                result[dcid] = entry['value'][:i18nname_end_index]
+            else:
+                if entry['value'].endswith('en'):
+                    name_in_en = entry['value'][:(len(entry['value']) -
+                                                  len('en') - 1)]
+        # With there isn't a name in required language code, falls back to 
+        # English name if exists.
```

Comment:
are we always guaranteed a name in english? i.e. if it's blank, it means we have no name at all for the node.

Reply:
There are 2408932 distinct subject_id that have i18n name, but only 1051262 have i18n name in english. 

Do you think that if there is no name in the required language, we shall fall back to english?
I think that it would be weird to show blank, if it does have english name. 
In that case, if it's blank, then there isn't a name in the required language code, neither in English.

Reply:
As far as I could tell, all the name is extracted from @en name. So if a place has a name, then it will also have an @en name. 
In that case, I don't think that we need to append the english name to the list of i18n names, since there will be an i18n english name. 
Just to confirm, there isn't another api call here. If we didn't find the required i18n name, we fall back to english name, which is already fetched in the same api call.